alns/select/MABSelector.py

```diff
@@ -0,0 +1,197 @@
+    from mabwiser.utils import Num
+except ModuleNotFoundError:
+    MABWISER_AVAILABLE = False
+
+
+def ops2arm(destroy_idx: int, repair_idx: int) -> str:
+    """
+    Converts a tuple of destroy and repair operator indices to an arm
+    string that can be passed to self._mab.
+
+    Examples
+    --------
+    >>> ops2arm(0, 1)
+    "0_1"
+    >>> ops2arm(12, 3)
+    "12_3"
+    """
+    return f"{destroy_idx}_{repair_idx}"
+
+
+def arm2ops(arm: str) -> Tuple[int, int]:
+    """
+    Converts an arm string returned from self._mab to a tuple of destroy
+    and repair operator indices.
+
+    Examples
+    --------
+    >>> arm2ops("0_1")
+    (0, 1)
+    >>> arm2ops("12_3")
+    (12, 3)
+    """
+    [destroy, repair] = arm.split("_")
+    return int(destroy), int(repair)
+
+
+class MABSelector(OperatorSelectionScheme):
+    """
+    A selector that uses any multi-armed-bandit algorithm from MABWiser.
+
+    This selector is a wrapper around the many multi-armed bandit algorithms
+    available in the `MABWiser <https://github.com/fidelity/mabwiser>`_
+    library. Since ALNS operator selection can be framed as a
+    multi-armed-bandit problem (where each [destroy, repair] operator pair is
+    a bandit arm), this wrapper allows you to use a variety of existing
+    multi-armed-bandit algorithms as operator selectors instead of
+    having to reimplement them.
+
+    Note that the supplied ``MAB`` object must be generated with the static
+    method ``make_arms``.
+
+    Parameters
+    ----------
+    scores
+        A list of four non-negative elements, representing the rewards when the
+        candidate solution results in a new global best (idx 0), is better than
+        the current solution (idx 1), the solution is accepted (idx 2), or
+        rejected (idx 3).
+    mab
+        A MABWiser MAB object that will be used to select the
+        (destroy, repair) operator pairs. The arms of the ``mab`` object must
+        be generated with the static method ``make_arms``.
+    num_destroy
+        Number of destroy operators.
+    num_repair
+        Number of repair operators.
+    op_coupling
+        Optional boolean matrix that indicates coupling between destroy and
+        repair operators. Entry (i, j) is True if destroy operator i can be
+        used together with repair operator j, and False otherwise.
+    context_extractor
+        Optional function that takes a ALNS State object and returns a context
+        vector for that state that can be passed to a contextual MABWiser
+        bandit. If the MAB algorithm supports it, this context will be used to
+        help predict the next (destroy, repair) combination.
+
+    References
+    ----------
+    .. [1] Emily Strong, Bernard Kleynhans, & Serdar Kadioglu (2021).
+           MABWiser: Parallelizable Contextual Multi-armed Bandits.
+           Int. J. Artif. Intell. Tools, 30(4), 2150021:1–2150021:19.
+    """
+
+    if not MABWISER_AVAILABLE:
```

Comment:
I think this needs to go in the init method, not in the class itself. I could be wrong, but I think this gets executed by the parser even when the user is not using `MABSelector`.

Reply:
Darn, I spent a bit puzzling through this and thought this approach might work, but you're correct. The problem with putting this in the constructor is that the type annotations on the constructor arguments get resolved before the function body runs, and this resolution will fail if the user doesn't have MABWiser installed.  Could we get away with stringifying the constructor arguments?
```diff
def __init__(
    ...
-    learning_policy: LearningPolicy,
+    learning_policy: "LearningPolicy",
    ...
)
```
It's unclear to me whether this still results in type checking with current mypy versions.

Reply:
I think that's OK for us!

Reply:
Pushed a fix. I tested with the following three scenarios:
1. Don't have mabwiser installed and don't try to use the `MABSelector` (works)
2. Don't have mabwiser installed and try to use the `MABSelector` (raises an exception)
3. Have mabwiser installed and try to use the `MABSelector` (works)